File: src/State/MainMenu.cpp
// Project includes
#include "../State/MainMenu.h"
#include "../Audio/Mixer.h"
#include "../Event/State.h"
#include "../functions.h"
#include "../Game/Game.h"
#include "../Graphics/Color.h"
#include "../Graphics/Renderer.h"
#include "../Input/Mouse.h"
#include "../Logger.h"
#include "../ResourceManager.h"
#include "../State/Credits.h"
#include "../State/LoadGame.h"
#include "../State/Movie.h"
#include "../State/NewGame.h"
#include "../State/SettingsMenu.h"
#include "../UI/Animation.h"
#include "../UI/Factory/ImageButtonFactory.h"
#include "../UI/Image.h"
#include "../UI/ImageButton.h"
#include "../UI/TextArea.h"

// Third-party includes

// stdlib
#include <sstream>

namespace Falltergeist
{
    namespace State
    {
        using ImageButtonType = UI::Factory::ImageButtonFactory::Type;
        using Point = Graphics::Point;

        MainMenu::MainMenu(std::shared_ptr<UI::IResourceManager> resourceManager, std::shared_ptr<ILogger> logger) : State()
        {
            this->resourceManager = std::move(resourceManager);
            this->logger = std::move(logger);
            imageButtonFactory = std::make_unique<UI::Factory::ImageButtonFactory>(this->resourceManager);
        }

        MainMenu::~MainMenu()
        {
            Game::Game::getInstance()->mixer()->stopMusic();
        }

        void MainMenu::init()
        {
            if (_initialized) {
                return;
            }
            State::init();

            setModal(true);
            setFullscreen(true);

            auto renderer = Game::Game::getInstance()->renderer();
            setPosition((renderer->size() - Point(640, 480)) / 2);

            addUI("background", resourceManager->getImage("art/intrface/mainmenu.frm"));

            // intro button
            auto introButton = addUI(imageButtonFactory->getByType(ImageButtonType::MENU_RED_CIRCLE, {30, 19}));
            introButton->mouseClickHandler().add(std::bind(&MainMenu::onIntroButtonClick, this, std::placeholders::_1));

            // new game button
            auto newGameButton = addUI(imageButtonFactory->getByType(ImageButtonType::MENU_RED_CIRCLE, {30, 19 + 41}));
            newGameButton->mouseClickHandler().add(std::bind(&MainMenu::onNewGameButtonClick, this, std::placeholders::_1));

            // load game button
            auto loadGameButton = addUI(imageButtonFactory->getByType(ImageButtonType::MENU_RED_CIRCLE, {30, 19 + 41 * 2}));
            loadGameButton->mouseClickHandler().add(std::bind(&MainMenu::onLoadGameButtonClick, this, std::placeholders::_1));

            // settings button
            auto settingsButton = addUI(imageButtonFactory->getByType(ImageButtonType::MENU_RED_CIRCLE, {30, 19 + 41 * 3}));
            settingsButton->mouseClickHandler().add(std::bind(&MainMenu::onSettingsButtonClick, this, std::placeholders::_1));

            // credits button
            auto creditsButton = addUI(imageButtonFactory->getByType(ImageButtonType::MENU_RED_CIRCLE, {30, 19 + 41 * 4}));
            creditsButton->mouseClickHandler().add(std::bind(&MainMenu::onCreditsButtonClick, this, std::placeholders::_1));

            // exit button
            auto exitButton = addUI(imageButtonFactory->getByType(ImageButtonType::MENU_RED_CIRCLE, {30, 19 + 41 * 5}));
            exitButton->mouseClickHandler().add(std::bind(&MainMenu::onExitButtonClick, this, std::placeholders::_1));

            auto font4 = ResourceManager::getInstance()->font("font4.aaf");
            Graphics::Color color = {0xb8, 0x9c, 0x28, 0xff};

            // "Intro" label
            auto introButtonLabel = std::make_shared<UI::TextArea>("INTRO", 50, 20);
            introButtonLabel->setFont(font4, color);
            introButtonLabel->setWidth(150);
            introButtonLabel->setHorizontalAlign(UI::TextArea::HorizontalAlign::CENTER);

            // "New Game" label
            auto newGameButtonLabel = std::make_shared<UI::TextArea>("NEW GAME", 50, 20 + 41);
            newGameButtonLabel->setFont(font4, color);
            newGameButtonLabel->setWidth(150);
            newGameButtonLabel->setHorizontalAlign(UI::TextArea::HorizontalAlign::CENTER);

            // "Load Game" label
            auto loadGameButtonLabel = std::make_shared<UI::TextArea>("LOAD GAME", 50, 20 + 41*2);
            loadGameButtonLabel->setFont(font4, color);
            loadGameButtonLabel->setWidth(150);
            loadGameButtonLabel->setHorizontalAlign(UI::TextArea::HorizontalAlign::CENTER);

            // "Options" label
            auto optionsButtonLabel = std::make_shared<UI::TextArea>("OPTIONS", 50, 20 + 41*3);
            optionsButtonLabel->setFont(font4, color);
            optionsButtonLabel->setWidth(150);
            optionsButtonLabel->setHorizontalAlign(UI::TextArea::HorizontalAlign::CENTER);

            // "Credits" label
            auto creditsButtonLabel = std::make_shared<UI::TextArea>("CREDITS", 50, 20 + 41*4);
            creditsButtonLabel->setFont(font4, color);
            creditsButtonLabel->setWidth(150);
            creditsButtonLabel->setHorizontalAlign(UI::TextArea::HorizontalAlign::CENTER);

            // "Exit" label
            auto exitButtonLabel = std::make_shared<UI::TextArea>("EXIT", 50, 20 + 41*5);
            exitButtonLabel->setFont(font4, color);
            exitButtonLabel->setWidth(150);
            exitButtonLabel->setHorizontalAlign(UI::TextArea::HorizontalAlign::CENTER);

            // Text areas
            addUI(introButtonLabel);
            addUI(newGameButtonLabel);
            addUI(loadGameButtonLabel);
            addUI(optionsButtonLabel);
            addUI(creditsButtonLabel);
            addUI(exitButtonLabel);
        }

        void MainMenu::doExit()
        {
            fadeDoneHandler().clear();
            fadeDoneHandler().add([this](Event::Event* event){ this->onExitStart(dynamic_cast<Event::State*>(event)); });
            Game::Game::getInstance()->renderer()->fadeOut(0,0,0,1000);
        }

        void MainMenu::doNewGame()
        {
            fadeDoneHandler().clear();
            fadeDoneHandler().add([this](Event::Event* event){ this->onNewGameStart(dynamic_cast<Event::State*>(event)); });
            Game::Game::getInstance()->renderer()->fadeOut(0,0,0,1000);
        }

        void MainMenu::doLoadGame()
        {
            fadeDoneHandler().clear();
            fadeDoneHandler().add([this](Event::Event* event){ this->onLoadGameStart(dynamic_cast<Event::State*>(event)); });
            Game::Game::getInstance()->renderer()->fadeOut(0,0,0,1000);
        }

        void MainMenu::doSettings()
        {
            Game::Game::getInstance()->pushState(new SettingsMenu(resourceManager));
        }

        void MainMenu::doIntro()
        {
            fadeDoneHandler().clear();
            fadeDoneHandler().add([this](Event::Event* event){ this->onIntroStart(dynamic_cast<Event::State*>(event)); });
            Game::Game::getInstance()->renderer()->fadeOut(0,0,0,1000);
        }

        void MainMenu::doCredits()
        {
            fadeDoneHandler().clear();
            fadeDoneHandler().add([this](Event::Event* event){ this->onCreditsStart(dynamic_cast<Event::State*>(event)); });
            Game::Game::getInstance()->renderer()->fadeOut(0,0,0,1000);
        }

        void MainMenu::onExitButtonClick(Event::Mouse* event)
        {
            doExit();
        }

        void MainMenu::onExitStart(Event::State* event)
        {
            fadeDoneHandler().clear();
            Game::Game::getInstance()->mixer()->stopMusic();
            Game::Game::getInstance()->quit();
        }

        void MainMenu::onNewGameButtonClick(Event::Mouse* event)
        {
            doNewGame();
        }

        void MainMenu::onNewGameStart(Event::State* event)
        {
            fadeDoneHandler().clear();
            Game::Game::getInstance()->pushState(new NewGame(resourceManager, logger));
        }

        void MainMenu::onLoadGameButtonClick(Event::Mouse* event)
        {
            doLoadGame();
        }

        void MainMenu::onLoadGameStart(Event::State* event)
        {
            fadeDoneHandler().clear();
            Game::Game::getInstance()->pushState(new LoadGame(resourceManager));
        }

        void MainMenu::onSettingsButtonClick(Event::Mouse* event)
        {
            doSettings();
        }

        void MainMenu::onIntroButtonClick(Event::Mouse* event)
        {
            doIntro();
        }

        void MainMenu::onIntroStart(Event::State* event)
        {
            fadeDoneHandler().clear();
            Game::Game::getInstance()->pushState(new Movie(17));
            Game::Game::getInstance()->pushState(new Movie(1));
        }

        void MainMenu::onCreditsButtonClick(Event::Mouse* event)
        {
            doCredits();
        }

        void MainMenu::onCreditsStart(Event::State* event)
        {
            fadeDoneHandler().clear();
            Game::Game::getInstance()->pushState(new Credits());
        }

        void MainMenu::onKeyDown(Event::Keyboard* event)
        {
            switch (event->keyCode())
            {
                case SDLK_e:
                case SDLK_ESCAPE:
                    doExit();
                    break;
                case SDLK_n:
                    doNewGame();
                    break;
                case SDLK_l:
                    doLoadGame();
                    break;
                case SDLK_i:
                    doIntro();
                    break;
                case SDLK_c:
                    doCredits();
                    break;
                case SDLK_o:
                    doSettings();
                    break;
            }
        }

        void MainMenu::onStateActivate(Event::State* event)
        {
            Game::Game::getInstance()->mouse()->setState(Input::Mouse::Cursor::BIG_ARROW);
            Game::Game::getInstance()->mixer()->playACMMusic("07desert.acm",true);
            Game::Game::getInstance()->renderer()->fadeIn(0,0,0,1000);
        }
    }
}

File: data/scripts/lua/state/MainMenu.lua
--
-- Copyright 2012-2015 Falltergeist Developers.
--
-- This file is part of Falltergeist.
--
-- Falltergeist is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- Falltergeist is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with Falltergeist.  If not, see <http://www.gnu.org/licenses/>.
--

-- requires section
local State = require("state.State")
local Image = require("ui.Image")
local ImageButton = require("ui.ImageButton")
local TextArea = require("ui.TextArea")
local input = require("input")


-- class/module table
local MainMenu = State:subclass "MainMenu"

function MainMenu:LittleRedButton(x, y)
    return ImageButton:new(self, "art/intrface/lilredup.frm", "art/intrface/lilreddn.frm", "sound/sfx/ib1lu1x1.acm", "sound/sfx/ib1p1xx1.acm",  x or 0, y or 0)
end

function MainMenu:BigMenuButton(x, y)
    return ImageButton:new(self, "art/intrface/menuup.frm", "art/intrface/menudown.frm", "sound/sfx/nmselec1.acm", "sound/sfx/nmselec0.acm",  x or 0, y or 0)
end

function MainMenu:BigLabel(textId, x, y)
    local ta = TextArea:new(self, game._t(enum.MSG_FILE.MISC, textId), x or 0, y or 0)
    ta:setFont("font4.aaf", 0xb89c28ff)
    ta.size = Size(150, 0)
    ta.horizontalAlign = enum.TEXT_ALIGN.CENTER
    return ta
end


function MainMenu:_doIntro()
    self:fadeOut(function()
        game.state.Movie.pushNew(17)
        game.state.Movie.pushNew(1)
    end)
end

function MainMenu:_doNewGame()
    self:fadeOut(function()
        game.state.NewGame.pushNew()
    end)
end

function MainMenu:_doLoadGame()
    self:fadeOut(function()
        game.state.LoadGame.pushNew()
    end)
end

function MainMenu:_doSettings()
    game.state.SettingsMenu.pushNew()
end

function MainMenu:_doCredits()
    self:fadeOut(function()
        game.state.Credits.pushNew()
    end)
end

function MainMenu:_doExit()
    self:fadeOut(function()
        game.mixer:stopMusic()
        game.quit()
    end)
end

function MainMenu:initialize()
    State.initialize(self)

    self.modal = true
    self.fullscreen = true

    game.mouse.cursor = enum.CURSOR.BIG_ARROW
    game.mouse.position = Point(400, 400)

    -- background
    Image:new(self, "art/intrface/mainmenu.frm")

    -- Buttons
    local introButton    = self:BigMenuButton(30, 19);
    local newGameButton  = self:BigMenuButton(30, 19 + 41);
    local loadGameButton = self:BigMenuButton(30, 19 + 41*2);
    local settingsButton = self:BigMenuButton(30, 19 + 41*3);
    local creditsButton  = self:BigMenuButton(30, 19 + 41*4);
    local exitButton     = self:BigMenuButton(30, 19 + 41*5);

    -- Labels
    self:BigLabel( 9, 50, 20) -- intro
    self:BigLabel(10, 50, 20 + 41) -- new game
    self:BigLabel(11, 50, 20 + 41*2) -- load game
    self:BigLabel(12, 50, 20 + 41*3) -- options
    self:BigLabel(13, 50, 20 + 41*4) -- credits
    self:BigLabel(14, 50, 20 + 41*5) -- exit

    -- Button click handlers
    local this = self
    introButton.obj:addEventHandler("mouseleftclick", function() this:_doIntro() end)
    newGameButton.obj:addEventHandler("mouseleftclick", function() this:_doNewGame() end)
    loadGameButton.obj:addEventHandler("mouseleftclick", function() this:_dodoLoadGame() end)
    settingsButton.obj:addEventHandler("mouseleftclick", function() this:_doSettings() end)
    creditsButton.obj:addEventHandler("mouseleftclick", function() this:_doCredits() end)
    exitButton.obj:addEventHandler("mouseleftclick", function() this:_doExit() end)
end

function MainMenu:onActivate()
    self:fadeIn()
    game.mixer:playACMMusic("07desert.acm", true)
end

function MainMenu:onKeyDown(event)
    local key = event.keyCode
    if key == input.SDLK_e or key == input.SDLK_ESCAPE then
        self:_doExit()
    elseif key == input.SDLK_n then
        self:_doNewGame()
    elseif key == input.SDLK_l then
        self:_doLoadGame()
    elseif key == input.SDLK_i then
        self:_doIntro()
    elseif key == input.SDLK_c then
        self:_doCredits()
    elseif key == input.SDLK_o then
        self:_doSettings()
    end
end

--function MainMenu:think()
--end

-- return class table
return MainMenu
